### ISSUE TITLE
Special parameter $# should be quoted in case numbers in IFS

### ISSUE THREAD

**Issue by krushia** [koalaman/shellcheck#3351]
#### For bugs with existing features

- Rule Id SC2086 (maybe?)
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
f() {
while test $# -gt 0
do
 shift
done
}

IFS=0
f a b c
```

This will cause test to error when $# is 0, as it becomes `test '' -gt 0`

#### Here's what shellcheck currently says:
Nothing

#### Here's what I wanted or expected to see:
```
while test $# -gt 0
           ^-- SC2086 (info): Double quote to prevent globbing and word splitting.
```

**Comment by krushia** [comment]
Same problem exists with `${#foo}` and `$((foo))`. It appears anything that looks like an integer has been erroneously whitelisted.

**Comment by serious-angel** [comment]
This is a mandatory and quite critical case to consider a warning for, indeed.
Thank you for the head-sup, @krushia ! 🛡️
